File: functions/lib/slots.js
// Slot occupancy math for SafeBeauty, extracted so it can be unit-tested without
// Firebase. A booking is not a single instant: a multi-service or group/wedding
// booking realistically takes several back-to-back slots. We approximate one slot
// per service (the salon's slotDurationMinutes is the granularity), so an
// appointment blocks `slotsCount` consecutive slots and can't be silently
// double-booked over. Legacy appointments (no slotsCount/services) fall back to a
// single slot, so nothing old breaks.
//
// ── Processing time ─────────────────────────────────────────────────────────
//
// Some services leave the stylist free in the middle of them. Colouring hair is
// roughly 45 minutes of application, 30 minutes while the colour develops, then
// 20 minutes of washing and styling — and during those 30 minutes the stylist
// can be cutting somebody else's hair. Without modelling that, the salon either
// wastes the gap or overbooks into it by hand, and the gap is on the most
// expensive service in the market.
//
// So a booking spans a number of slots, and only SOME of them are ones the
// stylist is working. `busyOffsets` names which; everything else here — conflict
// detection, the customer's slot picker — already works on sets of slot times,
// so leaving the free ones out of that set is the whole change.
//
// Quantised to whole slots, because a 30-minute gap cannot be expressed inside a
// 60-minute grid. A salon that wants finer control sets a shorter
// slotDurationMinutes. The rounding is deliberately asymmetric:
//
//     working time rounds UP, processing time rounds DOWN
//
// so that every rounding error costs the salon a little capacity rather than
// handing the same stylist to two customers at once. An error that loses money
// is recoverable; an error that double-books is a woman turned away at the door.

/** 0, 1, … n-1 */
function range(n) {
  const out = [];
  for (let i = 0; i < n; i++) out.push(i);
  return out;
}

// The slot offsets an appointment's stylist is actually working, relative to its
// start. Stored on the appointment at booking time rather than recomputed from
// the salon, so that changing a service's timing later cannot retroactively move
// a booking that is already in someone's calendar.
function busyOffsetsOf(appt, span) {
  const raw = appt && appt.busyOffsets;
  if (!Array.isArray(raw) || raw.length === 0) return range(span);
  const clean = raw
    .map(Number)
    .filter((n) => Number.isInteger(n) && n >= 0 && n < span);
  // An appointment that claims to occupy nothing would be invisible to the
  // conflict check and could be booked straight over. Fall back to the whole
  // span, which is what every appointment written before this existed means.
  return clean.length ? clean : range(span);
}

// The list of slot start-times a single appointment occupies — meaning the times
// its stylist is unavailable, which for a service with processing time is fewer
// than the times the client is in the chair.
function slotsForAppointment(appt, slotMinutes) {
  const start = Number(appt ? appt.appointmentDate : NaN);
  if (!Number.isFinite(start)) return [];
  const step = Math.max(1, Number(slotMinutes) || 30) * 60000;
  const span = Math.max(
    1,
    Number(appt.slotsCount) ||
      (Array.isArray(appt.services) ? appt.services.length : 0) ||
      1
  );
  return busyOffsetsOf(appt, span).map((i) => start + i * step);
}

// Expands a list of appointments into the flat set of taken slot times. Returns
// both the legacy `slots` (plain times) and the staff-aware `booked` shape the
// client uses to allow parallel bookings across a multi-staff salon.
function expandBooked(appointments, slotMinutes) {
  const slots = [];
  const booked = [];
  for (const a of appointments || []) {
    const staffId = String((a && a.staffId) || "");
    for (const t of slotsForAppointment(a, slotMinutes)) {
      slots.push(t);
      booked.push({ time: t, staffId });
    }
  }
  return { slots, booked };
}


/**
 * Where a booking's working time and its idle time fall.
 *
 * Returns the number of slots the booking spans and, of those, the offsets the
 * stylist is occupied. Services run back to back, so a colour followed by a
 * blow-dry puts the colour's development gap in the middle of the whole booking.
 *
 * [timings] is the salon's per-service breakdown — { activeBefore, processing,
 * activeAfter } in minutes. A service with no entry there falls back to
 * durationPerService and is treated as working throughout, which is exactly what
 * every salon means today, so nothing changes for one that never sets this.
 */
function serviceLayout(serviceNames, timings, durationPerService, slotMinutes) {
  const step = Math.max(1, Number(slotMinutes) || 30);
  const names = Array.isArray(serviceNames) ? serviceNames : [];
  const tim = timings && typeof timings === "object" ? timings : {};
  const durs = durationPerService && typeof durationPerService === "object" ? durationPerService : {};

  if (names.length === 0) return { span: 1, busyOffsets: [0] };

  const busy = [];
  let cursor = 0;
  for (const name of names) {
    const t = tim[name] || {};
    const before = Math.max(0, Number(t.activeBefore) || 0);
    const proc   = Math.max(0, Number(t.processing)   || 0);
    const after  = Math.max(0, Number(t.activeAfter)  || 0);

    let beforeSlots, procSlots, afterSlots;
    if (before + proc + after > 0) {
      // Working time up, idle time down — see the asymmetry note at the top.
      beforeSlots = Math.ceil(before / step);
      procSlots   = Math.floor(proc / step);
      afterSlots  = Math.ceil(after / step);
      // A service the stylist never works is not a service. If the rounding left
      // no working slots at all, keep one rather than produce a booking that
      // conflicts with nothing.
      if (beforeSlots + afterSlots === 0) beforeSlots = 1;
    } else {
      const d = Number(durs[name]);
      beforeSlots = Math.ceil((Number.isFinite(d) && d > 0 ? d : step) / step);
      procSlots = 0;
      afterSlots = 0;
    }

    for (let i = 0; i < beforeSlots; i++) busy.push(cursor + i);
    cursor += beforeSlots + procSlots;
    for (let i = 0; i < afterSlots; i++) busy.push(cursor + i);
    cursor += afterSlots;
  }

  const span = Math.max(1, cursor);
  return { span, busyOffsets: busy.length ? busy : [0] };
}

// True when a booking starting at [reqStart] for staff [staffId] would overlap
// any existing appointment. A different staffId is a different chair, so it never
// conflicts (parallel bookings across a multi-staff salon); a solo salon uses
// staffId "" for everything, so any overlap there conflicts. CANCELLED
// appointments are ignored, and [excludeId] (the appointment being rescheduled
// onto a new time) is skipped so it can't conflict with its own current slot.
// Each existing item may carry an `id`.
//
// [reqSpanOrOffsets] is either a slot count — every slot from the start is
// working time, which is what a booking without processing time means — or the
// explicit list of working offsets from serviceLayout. Both sides of the
// comparison are then just sets of times the stylist is busy, and a colour's
// development gap is simply not in either set.
function hasSlotConflict(existing, reqStart, reqSpanOrOffsets, staffId, slotMinutes, excludeId) {
  const start = Number(reqStart);
  if (!Number.isFinite(start)) return false;
  const step = Math.max(1, Number(slotMinutes) || 30) * 60000;
  const offsets = Array.isArray(reqSpanOrOffsets) && reqSpanOrOffsets.length
    ? reqSpanOrOffsets.map(Number).filter((n) => Number.isInteger(n) && n >= 0)
    : range(Math.max(1, Number(reqSpanOrOffsets) || 1));
  const wanted = new Set();
  for (const i of (offsets.length ? offsets : [0])) wanted.add(start + i * step);
  const wantStaff = String(staffId || "");
  for (const a of existing || []) {
    if (!a || a.status === "CANCELLED") continue;
    if (excludeId && a.id === excludeId) continue;
    if (String(a.staffId || "") !== wantStaff) continue;
    for (const t of slotsForAppointment(a, slotMinutes)) {
      if (wanted.has(t)) return true;
    }
  }
  return false;
}

module.exports = { slotsForAppointment, expandBooked, serviceLayout, hasSlotConflict };

File: ios/SafeBeautyCore/Sources/SafeBeautyCore/Booking/Slots.swift
import Foundation

/// The slot grid, ported from `functions/lib/slots.js` rule for rule.
///
/// This is the one piece of client logic where disagreeing with the server does
/// not produce a wrong screen — it produces two women in the same chair. The
/// server re-checks every booking with `hasSlotConflict` and rejects a
/// collision, so a client that computes a *wider* grid than the server merely
/// hides bookable times; a client that computes a *narrower* one offers times
/// that will be refused at payment, after she has chosen. Neither is
/// acceptable, so this matches exactly and is tested against vectors generated
/// from the server's own implementation.
public enum Slots {

    public struct Layout: Equatable, Sendable {
        /// How many slots of wall-clock the appointment spans.
        public let span: Int
        /// Which of those the chair is actually occupied for.
        public let busyOffsets: [Int]
    }

    /// Where a service's time goes: occupied, then unattended, then occupied.
    public struct Timing: Sendable, Equatable {
        public let activeBefore: Int   // minutes the stylist is with her
        public let processing: Int     // minutes the chair is free — dye developing
        public let activeAfter: Int

        public init(activeBefore: Int = 0, processing: Int = 0, activeAfter: Int = 0) {
            self.activeBefore = activeBefore
            self.processing = processing
            self.activeAfter = activeAfter
        }
    }

    /// Lay a list of services out across the grid.
    ///
    /// The asymmetry between `ceil` and `floor` is the whole point and is not a
    /// rounding preference. Active time is rounded UP, because a stylist who
    /// needs 35 minutes of a 30-minute slot occupies two. Processing time is
    /// rounded DOWN, because those minutes are when the chair is free and
    /// rounding them up would blank out a slot somebody else could have had.
    ///
    /// The guard that follows matters just as much: when a service is pure
    /// processing — dye that develops for an hour with no attended time — both
    /// counts round to zero and the booking would occupy the chair for no slots
    /// at all, so it could be double-booked against itself. One slot is always
    /// claimed.
    public static func serviceLayout(
        serviceNames: [String],
        timings: [String: Timing] = [:],
        durationPerService: [String: Int] = [:],
        slotMinutes: Int
    ) -> Layout {
        let step = max(1, slotMinutes)
        guard !serviceNames.isEmpty else { return Layout(span: 1, busyOffsets: [0]) }

        var busy: [Int] = []
        var cursor = 0

        for name in serviceNames {
            var beforeSlots: Int
            var procSlots: Int
            var afterSlots: Int

            let t = timings[name]
            let before = max(0, t?.activeBefore ?? 0)
            let proc = max(0, t?.processing ?? 0)
            let after = max(0, t?.activeAfter ?? 0)

            if before + proc + after > 0 {
                beforeSlots = Int((Double(before) / Double(step)).rounded(.up))
                procSlots = Int((Double(proc) / Double(step)).rounded(.down))
                afterSlots = Int((Double(after) / Double(step)).rounded(.up))
                // Pure processing would otherwise block nothing at all.
                if beforeSlots + afterSlots == 0 { beforeSlots = 1 }
            } else {
                // No timing breakdown: the whole duration is attended.
                let d = durationPerService[name]
                let minutes = (d.map { $0 > 0 } ?? false) ? d! : step
                beforeSlots = Int((Double(minutes) / Double(step)).rounded(.up))
                procSlots = 0
                afterSlots = 0
            }

            for i in 0..<max(0, beforeSlots) { busy.append(cursor + i) }
            cursor += beforeSlots + procSlots
            for i in 0..<max(0, afterSlots) { busy.append(cursor + i) }
            cursor += afterSlots
        }

        return Layout(span: max(1, cursor), busyOffsets: busy.isEmpty ? [0] : busy)
    }

    /// Which offsets an existing booking occupies.
    ///
    /// A booking with no usable `busyOffsets` blocks its whole span. That is
    /// the safe direction: an older booking written before the field existed
    /// must not be treated as blocking nothing, or it can be booked over.
    public static func busyOffsets(stored: [Int]?, span: Int) -> [Int] {
        guard let stored, !stored.isEmpty else { return Array(0..<span) }
        let clean = stored.filter { $0 >= 0 && $0 < span }
        return clean.isEmpty ? Array(0..<span) : clean
    }

    /// The wall-clock instants an existing booking occupies, in milliseconds.
    public static func slotsFor(
        startMillis: Int64,
        slotsCount: Int?,
        servicesCount: Int = 0,
        storedBusyOffsets: [Int]? = nil,
        slotMinutes: Int
    ) -> [Int64] {
        let step = Int64(max(1, slotMinutes)) * 60_000
        // slotsCount first, then how many services there are, then one. An
        // older booking has neither and still occupies its slot.
        let span = max(1, (slotsCount ?? 0) > 0 ? slotsCount! : (servicesCount > 0 ? servicesCount : 1))
        return busyOffsets(stored: storedBusyOffsets, span: span)
            .map { startMillis + Int64($0) * step }
    }

    /// Whether a requested booking collides with anything already there.
    ///
    /// Two rules that are easy to get subtly wrong:
    ///
    /// A party booking takes the whole salon, so it conflicts with every chair
    /// rather than only its own. That is checked from BOTH sides — an existing
    /// party blocks a new single booking, and a new party is blocked by any
    /// existing booking.
    ///
    /// A cancelled appointment never conflicts. It still exists as a document,
    /// and treating it as occupied would slowly make a busy salon unbookable.
    public static func hasConflict(
        existing: [Appointment],
        requestedStart: Int64,
        requestedOffsets: [Int],
        staffId: String,
        slotMinutes: Int,
        excludeId: String? = nil,
        requestIsParty: Bool = false
    ) -> Bool {
        let step = Int64(max(1, slotMinutes)) * 60_000
        let offsets = requestedOffsets.filter { $0 >= 0 }
        let wanted = Set((offsets.isEmpty ? [0] : offsets).map { requestedStart + Int64($0) * step })

        for appointment in existing {
            if appointment.status == .cancelled { continue }
            if let excludeId, appointment.id == excludeId { continue }

            let wholeSalon = requestIsParty || appointment.isParty
            if !wholeSalon && appointment.staffId != staffId { continue }

            let taken = slotsFor(
                startMillis: appointment.appointmentDate,
                slotsCount: appointment.slotsCount,
                servicesCount: appointment.services.count,
                storedBusyOffsets: appointment.busyOffsets.isEmpty ? nil : appointment.busyOffsets,
                slotMinutes: slotMinutes
            )
            if taken.contains(where: wanted.contains) { return true }
        }
        return false
    }
}
